keps/core/model-cross-worksapce-entitlements.md

```diff
@@ -0,0 +1,1004 @@
+implementation is intended to be sufficient for use-cases where the complexity of entitlements
+is low, the rate of change is sufficiently low and the incidence of strict entitlement rules
+is rare. This implementation will make use of two objects in the `entitlements.tenancy.kcp.io` group:
+
+```go
+type EntitlementPolicy struct {
+    metav1.TypeMeta `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    // Entitlements holds all the entitlements that this EntitlementPolicy
+    // groups. 
+    // +optional
+    Entitlements []kcpcorev1alpha1.Entitlement `json:"entitlements"`
+}
+
+type EntitlementPolicyBinding struct {
+    metav1.TypeMeta `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    // EntitlementPolicyRef is a reference to the entitlement policy
+    // that this binding applies to.
+    EntitlementPolicyRef EntitlementPolicyRef `json:"entitlementPolicyRef"`
+
+    // Children determines if this policy binding provides the entitlement just to the
+    // workspace in which it is created, or also to the child workspaces underneath that.
+    // Optional, defaults to false.
+    // +optional
+    Children boolean `json:"children,omitempty"`
+}
+
+type EntitlementPolicyRef struct {
+    // ClusterPath is the human-readable path to the logical cluster in which the entitlement
+    // policy is registered.
+    ClusterPath logicalcluster.Path `json:"clusterPath"`
+
+    // Name is the name of the entitlement policy.
+    Name string `json:"name"`
+}
+```
+
+A service provider with permissions to create `EntitlementPolicy` will do so using the
+normal `kcp` API:
+
+```
+> POST /clusters/<entitlement-provider-cluster>/apis/entitlements.tenancy.kcp.io/v1alpha1/entitlementpolicies
+> {
+>     "kind": "EntitlementPolicy",
+>     "apiVersion": "entitlements.tenancy.kcp.io/v1alpha1",
+>     "metadata": {
+>         "name": "something-meaningful"
+>     },
+>     "entitlements": [
+>         {
+>             "clusterName": "33bab531",
+>             "kind": "WidgetAccess",
+>             "apiVersion": "acme.corp/v1",
+>             "spec": {
+>                 "something": {
+>                     "meaningful": "here"
+>                 }
+>             }
+>         }
+>     ]
+> }
+201 CREATED
+{
+    "kind": "EntitlementPolicy",
+    "apiVersion": "entitlements.tenancy.kcp.io/v1alpha1",
+    "metadata": {
+        "name": "something-meaningful"
+    },
+    "entitlements": [
+        {
+            "clusterName": "33bab531",
+            "kind": "WidgetAccess",
+            "apiVersion": "acme.corp/v1",
+            "spec": {
+                "something": {
+                    "meaningful": "here"
+                }
+            }
+        }
+    ]
+}
+```
+
+However, to entitle tenants, a per-tenant endpoint is used:
+
+```
+> POST /services/entitlements/clusters/<entitlement-consumer-cluster>/apis/entitlements.tenancy.kcp.io/v1alpha1/entitlementpolicybindings
```

Comment:
> maybe intentional: this means that this POST goes to the same shard as the APIExport workspace?

I expect this to be unrelated from APIExports, actually - we already have three distinct use-cases in our current code-base and I imagine Edge MC will have ones as well. 

> And that the VW is clever enough and has permissions to route the request to the right shard where the tenant workspace lives? This might be possible to do, but need deeper design.

Orthogonally from that, though, yes, if we expect a user to interact with one endpoint and we need the output objects to go to the right shard, we need that routing logic. Doesn't that problem exist already today with any other view?

> Also: where does an entitler know this URL from? It must be set somewhere, e.g. on an APIExport.

I was expecting something like your `ThingEntitlementClass` idea, good call-out that this is underspecified.

Reply:
> Doesn't that problem exist already today with any other view?

Am not aware of any such case. It's not trivial I fear, basically depends on the workspace index.

Reply:
One thought: would add the provider cluster and policy name in here:

```
POST /services/entitlements/clusters/<provider>/<policy-name>/clusters/<entitlement-consumer-cluster>/apis/entitlements.tenancy.kcp.io/v1alpha1/entitlementpolicybindings
```

That way we could authz the access easily against the policy object.